source/adapters/level_zero/image.cpp

```diff
@@ -36,75 +586,151 @@ urBindlessImagesUnsampledImageHandleDestroyExp(ur_context_handle_t hContext,
+      UR_CALL(createUrMemFromZeImage(hContext, ZeImage, /*OwnZeMemHandle*/ true,
+                                     ZeImageDesc, phMem));
+    } else {
+      ZeImage = UrImage->ZeImage;
+      *phMem = nullptr;
+    }
+  } else {
+    ze_image_pitched_exp_desc_t PitchedDesc;
+    PitchedDesc.stype = ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC;
+    PitchedDesc.pNext = nullptr;
+    PitchedDesc.ptr = hImageMem;
+
+    ze_image_bindless_exp_desc_t BindlessDesc;
+    BindlessDesc.stype = ZE_STRUCTURE_TYPE_BINDLESS_IMAGE_EXP_DESC;
+    BindlessDesc.pNext = &PitchedDesc;
+    BindlessDesc.flags = ZE_IMAGE_BINDLESS_EXP_FLAG_BINDLESS;
+
+    ZeImageDesc.pNext = &BindlessDesc;
+
+    ZE2UR_CALL(zeImageCreate, (hContext->ZeContext, hDevice->ZeDevice,
+                               &ZeImageDesc, &ZeImage));
+    ZE2UR_CALL(zeContextMakeImageResident,
+               (hContext->ZeContext, hDevice->ZeDevice, ZeImage));
+    UR_CALL(createUrMemFromZeImage(hContext, ZeImage, /*OwnZeMemHandle*/ true,
+                                   ZeImageDesc, phMem));
+  }
+
+  static std::once_flag InitFlag;
+  std::call_once(InitFlag, [&]() {
+    ze_driver_handle_t DriverHandle = hContext->getPlatform()->ZeDriver;
+    auto Result = zeDriverGetExtensionFunctionAddress(
+        DriverHandle, "zeImageGetDeviceOffsetExp",
+        (void **)&zeImageGetDeviceOffsetExpFunctionPtr);
+    if (Result != ZE_RESULT_SUCCESS)
+      urPrint("zeDriverGetExtensionFunctionAddress zeImageGetDeviceOffsetExp "
+              "failed, err = %d\n",
+              Result);
+  });
+  if (!zeImageGetDeviceOffsetExpFunctionPtr)
+    return UR_RESULT_ERROR_INVALID_OPERATION;
+
+  uint64_t DeviceOffset{};
+  ZE2UR_CALL(zeImageGetDeviceOffsetExpFunctionPtr, (ZeImage, &DeviceOffset));
+  *phImage = reinterpret_cast<ur_exp_image_handle_t>(DeviceOffset);
+
+  return UR_RESULT_SUCCESS;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL urBindlessImagesSampledImageCreateExp(
     ur_context_handle_t hContext, ur_device_handle_t hDevice,
     ur_exp_image_mem_handle_t hImageMem, const ur_image_format_t *pImageFormat,
     const ur_image_desc_t *pImageDesc, ur_sampler_handle_t hSampler,
     ur_mem_handle_t *phMem, ur_exp_image_handle_t *phImage) {
-  std::ignore = hContext;
-  std::ignore = hDevice;
-  std::ignore = hImageMem;
-  std::ignore = pImageFormat;
-  std::ignore = pImageDesc;
-  std::ignore = hSampler;
-  std::ignore = phMem;
-  std::ignore = phImage;
-  urPrint("[UR][L0] %s function not implemented!\n", __FUNCTION__);
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+
+  UR_CALL(urBindlessImagesUnsampledImageCreateExp(
+      hContext, hDevice, hImageMem, pImageFormat, pImageDesc, phMem, phImage));
+
+  struct combined_sampled_image_handle {
+    uint64_t raw_image_handle;
+    uint64_t raw_sampler_handle;
+  };
+  combined_sampled_image_handle *sampledImageHandle =
+      reinterpret_cast<combined_sampled_image_handle *>(phImage);
```

Comment:
there is no mechanism to prevent wrong casts ?
 
we rely on application calling  XXXSampledImageXXX() function with only Sampled Images

and UnsampledImage - with non sampled images? 

pasign Unsampled image handle to this function will crash  / or be UB

Reply:
ur_exp_image_handle_t size is 64bit for unsampled image and 128bit for sampled image for L0 adaptor.
Unfortunately size info isn't passed to this function.
In the long run the ideal solution is likely to support 64bit sampled image handle. For instance, can we always set bindless sampler offset to be the same as bindless image offset when zeImageCreate accepts a sampler argument in the future like cuda api cuTexObjectCreate? Then only one 64bit handle is enough for sampled image.